R/tm_rmarkdown.R

```diff
@@ -0,0 +1,264 @@
+    post_output = args$post_output
+  )
+}
+
+# Server function for the rmarkdown module
+srv_rmarkdown <- function(id, data, rmd_file, allow_download, extra_transform) {
+  checkmate::assert_class(data, "reactive")
+  checkmate::assert_class(isolate(data()), "teal_data")
+  moduleServer(id, function(input, output, session) {
+    if (allow_download) {
+      output$download_rmd <- downloadHandler(
+        filename = function() basename(rmd_file),
+        content = function(file) file.copy(rmd_file, file),
+        contentType = "text/plain"
+      )
+    }
+
+    pre_decorated_q_r <- reactive({
+      data_q <- req(data())
+      teal.reporter::teal_card(data_q) <- c(
+        teal.reporter::teal_card(data_q),
+        teal.reporter::teal_card("## Module's output(s)")
+      )
+      eval_code(
+        data_q,
+        sprintf(
+          "params <- list(%s)",
+          toString(sprintf("%1$s = %1$s", sapply(names(data_q), as.name)))
+        )
+      )
+    })
+
+    q_r <- data_with_output_decorated <- teal::srv_transform_teal_data(
+      "extra_transform",
+      data = pre_decorated_q_r,
+      transformators = extra_transform
+    )
+
+    temp_dir <- tempdir()
+    temp_rmd <- tempfile(tmpdir = temp_dir, fileext = ".Rmd")
+    file.copy(rmd_file, temp_rmd) # Use a copy of the Rmd file to avoid modifying the original
+
+    rendered_path_r <- reactive({
+      datasets <- req(q_r()) # Ensure data is available
+      tryCatch(
+        {
+          rmarkdown::render(
+            temp_rmd,
+            output_format = rmarkdown::md_document(
+              variant = "gfm",
+              toc = TRUE,
+              preserve_yaml = TRUE
+            ),
+            params = datasets[["params"]],
+            envir = new.env(parent = globalenv()),
+            quiet = TRUE,
+            runtime = "static"
+          )
+        },
+        error = function(e) {
+          warning("Error rendering RMD file: ", e$message) # verbose error in logs
+          e
+        }
+      )
+    })
+
+    rendered_html_r <- reactive({
+      output_path <- req(rendered_path_r())
+      validate(
+        need(inherits(output_path, "character"), "Error rendering RMD file. Please contact the app developer.")
+      )
+      htmltools::includeMarkdown(output_path)
+    })
+
+    output$rmd_output <- renderUI(rendered_html_r())
+
+    reactive({
+      out_data <- q_r()
+
+      if (allow_download) {
+        out_data <- eval_code(
+          q_r(),
+          paste(
+            sep = "\n",
+            sprintf("## R Markdown contents are generated from file, please download it from the module UI."),
+            sprintf("# rmarkdown::render(%s, params = params)", shQuote(basename(rmd_file), type = "cmd"))
+          )
+        )
+        out_data@verified <- FALSE # manual change verified status as code is being injected
+      }
+
+      teal.reporter::teal_card(out_data) <- c(
+        teal.reporter::teal_card(out_data),
+        rendered_html_r()
```

Comment:
markdown fits here better. We can also split it into `list()` suitable with `teal_card()`

Reply:
> We could keep it as a single markdown object

I think its good to have a single markdown string because anyway at the end (before download) we will concatenate everything to the single file. Only editor will show one combined text-field for this element (i think it is not a big issue so far) 

> the current approach was meant to reduce duplicate rendering

This is a fair reason. I feel (not a strong opinion) that if you render markdown using `rmarkdown::render` and `includeMarkdown` the speed should be similar to `rmarkdown::render` to html. `rmarkdown` renders in two stages to html, first by converting `Rmd -> md` and then pandoc renders from `md -> html`.

The biggest issue with including html is that html can't be converted to pdf, otf etc.